### PR TITLE
Redesigned routing and anchoring

### PR DESCRIPTION
**Pull request by JanKoehnlein**
Lots of changes:
- introduced manhattan routing, polyline routing and their common part linear routing
- added a registry for anchor computers
- introduced a routing module
- unified and integrated move, move handles and reconnect commands
- added self edge routing
- fixed undo/redo of all commands
- converted Routable interface into SRoutableElement classs
- collected routing related classes in routing/model
- avoid showing hovers when mouse enters while button is pressed

Missing features will be covered in separate issue.

### CODE REVIEW

File: src/features/routing/routing.ts [hunk 1]
```diff
@@ -14,17 +14,44 @@
...
+
 export interface IEdgeRouter {
-    route(edge: Routable): RoutedPoint[]
+
+    readonly kind: string;
+
+    /**
+     * Calculates the route of the given edge.
+     *
+     * @param edge
```
Comment (spoenemann) [inline src/features/routing/routing.ts]: Minor: remove undocumented `@param` tag. Same for other methods in this interface.

File: src/features/decoration/decoration-placer.ts [hunk 2]
```diff
@@ -14,18 +14,22 @@
...
 import { isDecoration, Decoration } from "./model";
 import { setAttr } from "../../base/views/vnode-utils";
 import { Point, ORIGIN_POINT } from "../../utils/geometry";
 import { isSizeable } from "../bounds/model";
-import { isRoutable } from "../routing/model";
+import { SRoutableElement } from "../routing/model";
+import { EdgeRouterRegistry } from "../routing/routing";
 
 @injectable()
 export class DecorationPlacer implements IVNodeDecorator {
```
[the remark this answers is not in the document]
Comment (JanKoehnlein) [reply src/features/decoration/decoration-placer.ts]: Filed #56

File: src/features/hover/hover.ts [hunk 3]
```diff
@@ -129,11 +129,24 @@ export interface HoverState {
 }
 
 export abstract class AbstractHoverMouseListener extends MouseListener {
+
+    protected mouseIsDown: boolean;
```
[the remark this answers is not in the document]
Comment (JanKoehnlein) [reply src/features/hover/hover.ts]: Having read the (docs)[https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/button]
> This property only guarantees to indicate which buttons are pressed during events caused by pressing or releasing one or multiple buttons. As such, it is not reliable for events such as mouseenter, mouseleave, mouseover, mouseout or mousemove.
I decided to got the other way.

**Comment by JanKoehnlein** [comment]
Thanks a lot for the review. Here's to your points:
> Class Diagram Example: Create a polyline route, then drag an anchor => route is deleted.

That's intended. At least for manhattan routing it doesn't make sense to keep the existing route when reconnecting an edge. To be discussed.

> Class Diagram Example: Start dragging the polyline anchor at "Bar", move it below the node, wait a bit until the hover popup appears _below the mouse cursor_, release mouse button => edge is left with a dangling anchor. 

Was a missing guard in `HoverMouseListener#mouseMove()`. Fixed

> Circles example: Performance is very bad when moving nodes manually.
> Circles example: Strange red artifacts appear in the top left corner when moving nodes manually.

Actually both are the same problem: We updated the edge handles when source/target are moved, regardless of whether the edge was editable or not. The red thingies and poor performance were a follow up as we haven't registered views for the routing handles. Fixed

File: src/features/routing/routing.ts (hunk 1 above)
Comment (JanKoehnlein) [reply src/features/routing/routing.ts]: Done

**Pull request merged**